src/b-splines/prefiltering.jl

```diff
@@ -53,6 +62,53 @@ function prefilter!(
     ret
 end
 
+function diffop(::Type{TWeights}, n::Int, k::Int) where {TWeights}
+    D = spdiagm(0 => ones(TWeights,n))
+    for i in 1:k
+        D = diff(D; dims=1)
+    end
+    ## TODO: Normalize by n?
+    D' * D
+end
+diffop(n::Int, k::Int) = diffop(Float64, n, k)
+### TODO: add compiled constructor for most common operators of order k=1,2
+#
+#function diffop(::Type{TWeights}, n::Int, k::Int) where {TWeights}
+#    D1 = spdiagm(0 => -ones(TWeights,n-1), 1 => ones(TWeights,n-1))[1:end-1, :]
+#    D = D1
+#    for i in 1:k-1
+#        D = diff(D; dims=1)
+##        D = D1[1+i:end, 1+i:end] * D
+##        lmul!(D1[1+i:end, 1+i:end], D)
+#    end
+#
+#    D' * D
+#end
+
+function prefilter!(
+    ::Type{TWeights}, ret::TCoefs, it::BSpline, λ::Real, k::Int
+    ) where {TWeights,TCoefs<:AbstractArray}
+    local buf, shape, retrs
+    sz = size(ret)
+    first = true
+    if ndims(ret) > 1
+        @warn "Smooth BSpline only available for Vectors, fallback to non-smooth"
```

Comment:
How about restricting `interpolate` to `AbstractVector` when `λ` and `k` are supplied?

Reply:
It should work for higher dimensions too, it is just not implemented yet.
When implemented, it will be easier to change then.